### PR TITLE
[for consideration] Hamster Repository

### PR DESCRIPTION
**Pull request by no-reply**
I think this is as ready as a simple implementation is going to get. 

We should decide whether this will suit as a replacement for the Hash-based implementation, or whether they need to co-exist.

Here's the state of things on my machine:

__Hamster:__

```
Calculating -------------------------------------
       query_pattern   646.000  i/100ms
       insert/delete     1.310k i/100ms
-------------------------------------------------
       query_pattern      6.621k (± 1.3%) i/s -     66.538k
       insert/delete     13.252k (± 1.0%) i/s -    133.620k
```

__Hash:__

```
Calculating -------------------------------------
       query_pattern   775.000  i/100ms
       insert/delete     3.285k i/100ms
-------------------------------------------------
       query_pattern      7.810k (± 6.6%) i/s -     78.275k
       insert/delete     34.252k (± 3.4%) i/s -    344.925k
```

All benchmarks are running for 10 seconds. 

`insert/delete` simply adds a statement (with a unique object) and deletes it.  It's possible we can optimize this more in the Hamster repository, but I'll leave that in the air for the moment.

I haven't looked closely at memory usage, yet.

Whatever we decide to do, I'll keep working on transactions.

### CODE REVIEW

**Comment by no-reply** [comment]
I'm stepping away from this for the night.  Feel free to hack on it if you want.  I might try to build out some better general repository benchmarks tomorrow.  Maybe a separate project (`rdf/benchmark`) that uses the [Berlin data generator](http://wifo5-03.informatik.uni-mannheim.de/bizer/berlinsparqlbenchmark/spec/BenchmarkRules/index.html#datagenerator) would be useful.

**Comment by gkellogg** [comment]
:+1:

**Comment by gkellogg** [comment]
`Repository::Implementation#has_statement?` is broken compared to the default Hash implementation because objects are maintained with a `Hamster::Set` rather than an `Array`. There is a different between `(Hamster::)Set`#include? and `Array#include?` in that `Array#include?` uses == for comparison, and Set looks for the same object (similar to a Hash key). Query depends on the use of == to do appropriate L2V mapping.

In the particular case I'm tracking in ld-patch, `"Cheers"@en-UK == "Cheers"@en-uk` but they are not equivalent. `Hamster::Set#include?` is implemented basically using a Hash.

Fixing this probably means using `#any? {|e| e.eql?(o)}` instead, which won't be quite as performant, but will preserve our needed semantics.

**Comment by no-reply** [comment]
Hmmmm... This problem could be worse than just #has_statement. We should
make sure there aren't scenarios where we end up with duplicate statements.

**Comment by gkellogg** [comment]
The fact that objects are in a Set should handle this, depending on how Hamsteer::Set#insert looks for equality.. It shouldn't be a problem but is worth investigating.

A more performing implementation would just store normalized literalls, or a dictionary index, but wouldn't allow for fuller literal equivalence. IMO, we need that for a default implementation.

**Comment by no-reply** [comment]
I'm understanding this better in the daylight hours.

`Hamster::Trie` uses `#hash` + `#eql?` to define its keyspace, so that should be fine for our purposes.

The issue you're running into seems to be isolated to the case sensitivity of language tags, and blank nodes. Otherwise, our definition of equality seems to work here. Dealing with the two exceptional cases:

__Language Tags:__

With the changes in f4c714b, while we get a term-equality sensitive `#has_statement?` we lose the ability to store the different language tags:

```ruby
repo = RDF::Repository.new

repo.insert RDF::Statement(RDF::URI('http://ex.org/1'), RDF::URI('http://ex.org/p'), RDF::Literal('abc', language: 'en-UK'))

repo.insert RDF::Statement(RDF::URI('http://ex.org/1'), RDF::URI('http://ex.org/p'), RDF::Literal('abc', language: 'en-uk'))

repo.count # => 1
```

We also trade constant time for (worst case) linear time access on `o` Sets ( `Hamster::Enumerable#any?` is implemented with `#each`). This feels like a nasty trade-off.

RDF gives us the option to universally downcase language tags [Lexical representations of language tags may be converted to lower case.](https://www.w3.org/TR/rdf11-concepts/#section-Graph-Literal). Doing so is an option, though I agree that ideally we would support the distinction between the lexical space and value space. This is in better keeping with our goal of being a quality reference implementation.

Another alternative would be for `#has_statement?` to always check full [literal term equality](https://www.w3.org/TR/rdf11-concepts/#dfn-literal-term-equality). This strikes me as the best behavior for the default implementation. Repositories that inline literals into value space can support coarser notions of equality as required--but should do so consistently.

__Blank Nodes:__

I think the problem here is more troubling. Our definition of `RDF::Node` equality gives us:

```ruby
RDF::Node.new('blah') == RDF::Node.new('blah') # => true
RDF::Node.new('blah').hash == RDF::Node.new('blah').hash # => true
RDF::Node.new('blah').eql? RDF::Node.new('blah') # => false
```

and, accordingly:

```ruby
repo = RDF::Repository.new

repo.insert RDF::Statement(RDF::Node.new('blah'), RDF::URI('http://ex.org/p'), RDF::Literal('abc', language: 'en-uk'))

repo.insert RDF::Statement(RDF::Node.new('blah'), RDF::URI('http://ex.org/p'), RDF::Literal('abc', language: 'en-uk'))

repo.count # => 2
repo.dump :ntriples
# => "_:blah <http://ex.org/p> \"abc\"@en-uk .\n_:blah <http://ex.org/p> \"abc\"@en-uk .\n"
```
Changing this means either:

  - Redefining `RDF::Node` equality to be in line with the other definitions; `#eql?` and `#hash` are both `true` whenever term equality will be fulfilled in serialization.
  - Accepting linear time access on `g` and `s` Hashes, as well as `o` Sets, for insert/delete/query in addition to`#has_statement?`.
  - Treating two separate `RDF::Node`s with the same IDs as fully distinct internally, and generating new IDs for them when serializing to avoid duplicate statements.

I think my preferred solution is the first option, but I'm open to the third.

Note that this issue (and constant/linear cost) applies equally to the RubyHash implementation; I'll submit an issue and a failing test in a bit.

**Comment by no-reply** [comment]
@bendiken may be interested in chiming in on https://github.com/ruby-rdf/rdf/pull/260#issuecomment-172360157.

I would certainly be glad for his input.

**Comment by bendiken** [comment]
Nice work, @no-reply! I can't dig into this right this instant, but will have a look later tonight.

**Comment by gkellogg** [comment]
> Hamster::Trie uses #hash + #eql? to define its keyspace, so that should be fine for our purposes.

The two statements are #eql? to each other, but this fails the `@trie.has_key?` test used to implement `Set#include?`. My proposed change `any? {|e| e.eql?(o)}` works at the expense of being more expensive.

> RDF gives us the option to universally downcase language tags ("Lexical representations of language tags may be converted to lower case.").

This is probably what we should do in `Literal`, which solves this specific issue.

**Blank Nodes**
> Changing this means either:

> * Redefining RDF::Node equality to be in line with the other definitions; #eql? and #hash are both true whenever term equality will be fulfilled in serialization.
> * Accepting linear time access on g and s Hashes, as well as o Sets, for insert/delete/query in addition to#has_statement?.
> * Treating two separate RDF::Nodes with the same IDs as fully distinct internally, and generating new IDs for them when serializing to avoid duplicate statements.

>I think my preferred solution is the first option, but I'm open to the third.

> Note that this issue (and constant/linear cost) applies equally to the RubyHash implementation; I'll submit an issue and a failing test in a bit.

No, BNode equality is specifically implemented this way as required by various SPARQL tests. Two nodes with the same identifier are not the same node, except when they're parsed in the same context. This requires most parsers to keep a local memory have allocated BNodes so that the same node is reused. I don't require specifically why `#==` is a looser interpretation, but the comment in RDF::Node describes the case. Changing this will cause SPARQL tests to break.

Depending on how `Hamster::Hash#has_key?` is implemented, two separate objects which are `eql?` to each other (due to s dup) should be the same node, and `#has_statement?` should verify this. Some tests are required:

```ruby
s = RDF::Statement(:a, :b, :c)
g = RDF::Graph.new
g << s
g.has_statement?(s) #=> true
s1 = s.dup
s1.subject = s1.subject.dup
g.has_statement?(s1) #=> true
```

This appears to work properly

```ruby
s1.subject = RDF::Node(:a)
g.has_statement?(s1) #=> false
```

This also works okay, so I think we're okay on BNodes.

I'll work on an update to `RDF::Literal` to map language tags to lower case and see what that might break, and also revert the change to Repository that uses `all?`.

File: lib/rdf/repository.rb [hunk 1]
```diff
@@ -285,14 +304,10 @@ def has_statement?(statement)
...
-          @data.dup.each do |g, ss|
-            ss.dup.each do |s, ps|
-              ps.dup.each do |p, os|
-                os.dup.each do |o|
+          @data.each do |g, ss|
+            ss.each do |s, ps|
+              ps.each do |p, os|
+                os.each do |o|
                   # FIXME: yield has better performance, but broken in MRI 2.2: See https://bugs.ruby-lang.org/issues/11451.
                   block.call(RDF::Statement.new(s, p, o, graph_name: g.equal?(DEFAULT_GRAPH) ? nil : g))
```
Comment (no-reply) [inline lib/rdf/repository.rb]: This is patched in 2.2.4 (and in 2.3.0 and above). See: Mon Nov 23 00:19:51 2015  Koichi Sasada  <ko1@atdot.net>

Should we switch back to `yield` and declare `2.2` versions before `2.2.4` unsupported?

**Comment by no-reply** [comment]
> The two statements are #eql? to each other, but this fails the @trie.has_key? test used to implement Set#include?. 

Yes. We're on the same page here.

> My proposed change any? {|e| e.eql?(o)} works at the expense of being more expensive.

I think the expense is bearable (though not ideal), since in most cases we shouldn't have very large lists of objects in the leaves. I'm more concerned about the awkward handling of case sensitivity. I think should either fully support case sensitivity in language tags (and I'm still in favor of this), or force downcase everything per _RDF Concepts_.

> >RDF gives us the option to universally downcase language tags ("Lexical representations of language tags may be converted to lower case.").
>
>This is probably what we should do in Literal, which solves this specific issue.

It's also unnecessary if we use strict literal equality in `#has_statement?`, and require clients looking to compare value equality to query through other channels. What would be the cost of this on the ld-patch side? Does it have SPARQL implications?

**Comment by gkellogg** [comment]
I think that someone why expects two literals that are the same except for the case of their language tag to be different will be disappointed anyway. The only difference would be on statement count, as all equivalence tests downcase the language for comparison anyway. I submitted a commit which did this and reverted the change to `#has_statement?` which lead to no other errors found either in RDF.rb or SPARQL.

That change also satisfies the tests for ld-patch; I think it's probably the way to go. If we were to do anything else, we might consider transforming to the proper language case (from a BCP47 perspective) when serializing (`#to_base`), but that conveys no different semantic value, and is purely aesthetic. Most other RDF implementations will just stick to lower-case, IIRC.

> At all times, language tags and their subtags, including private use
>   and extensions, are to be treated as case insensitive: there exist
>   conventions for the capitalization of some of the subtags, but these
>   MUST NOT be taken to carry meaning.

>   Thus, the tag "mn-Cyrl-MN" is not distinct from "MN-cYRL-mn" or "mN-
>   cYrL-Mn" (or any other combination), and each of these variations
>   conveys the same meaning: Mongolian written in the Cyrillic script as
>   used in Mongolia.

**Comment by no-reply** [comment]
> Depending on how `Hamster::Hash#has_key?` is implemented, two separate objects which are `eql?` to each other (due to s dup) should be the same node, and `#has_statement?` should verify this.

`#has_key?` is implemented with `Trie#get` (`!!get(key)`).  Equality in the `Trie` keyspace requires both `#eql?` and `#hash`.  This is also true of the Ruby Hash implementation, and the bug I'm pointing to applies to the old implementation, too.

The tests at https://github.com/ruby-rdf/rdf-spec/compare/feature/equality#diff-615339bfa07a4d62631f2ad337fb3898R150 and https://github.com/ruby-rdf/rdf-spec/commit/b743073f86d330cfda2a52642e6568f3c2227e6e#diff-8f2db15096fa69057d25ca99e3b367e4R108 offer one interpretation of the bug. I understand that fixing these specs would require a realignment of the various places where node equality lives, and that it may not be the best solution.

Still, we really need to either fix those specs as written, or dedup equal ids from distinct nodes when serializing. The ntriples output quoted above is definitely a bug.

File: lib/rdf/repository.rb (hunk 1 above)
Comment (gkellogg) [inline lib/rdf/repository.rb]: Yes, I think we can revert this. I saw it the other day too. I don't think we need to make any explicit statement about earlier patch revisions of MRI to be unsupported.

**Comment by no-reply** [comment]
:+1: to the language downcase solution.

**Comment by gkellogg** [comment]
if subject.mutable?
          subject.insert([RDF::Node.new('blah'), RDF::URI('p'), 'o'])
          expect { subject.delete([RDF::Node.new('blah'), RDF::URI('p'), 'o']) }
            .to change { subject.count }.by(-1)
        end
      end

This is a bug, and should not cause the graph to remove the statement, as it's a different node. It should be changed to `not_to`.

    it "should not insert a statement with duplicate node" do
      if subject.writable?
        statement.subject = RDF::Node.new('blah')
        subject.insert(statement)

        statement.subject = RDF::Node.new('blah')
        subject.insert(statement)
        expect(subject.count).to eq 1
      end
    end

This is also a bug, as it is a different node. It would be true if it were `statement.subject = statement.subject.dup`.

**Comment by no-reply** [comment]
@gkellogg I feel we're talking across each other. 

I'm fine with changing those specs as you say. I'm not fine with [this](https://github.com/ruby-rdf/rdf/pull/260#issuecomment-172360157):

```ruby
repo = RDF::Repository.new

repo.insert RDF::Statement(RDF::Node.new('blah'), RDF::URI('http://ex.org/p'), RDF::Literal('abc', language: 'en-uk'))

repo.insert RDF::Statement(RDF::Node.new('blah'), RDF::URI('http://ex.org/p'), RDF::Literal('abc', language: 'en-uk'))

repo.count # => 2
repo.dump :ntriples
# => "_:blah <http://ex.org/p> \"abc\"@en-uk .
            _:blah <http://ex.org/p> \"abc\"@en-uk .\n"
```

This repository says it contains two statements, but the serialization contains the same statement twice. Other serializations that use the `#id` will suffer the same problem.

**Comment by gkellogg** [comment]
See https://github.com/ruby-rdf/rdf/issues/261#issuecomment-172379662 for a discussion of serialization issues with BNode identifiers.

**Pull request merged**